hazelcast/src/main/java/com/hazelcast/nio/tcp/ProtocolDecoder.java

```diff
@@ -0,0 +1,152 @@
+
+    public ProtocolDecoder(IOService ioService, ProtocolEncoder protocolEncoder) {
+        this.ioService = ioService;
+        this.protocolEncoder = protocolEncoder;
+        this.props = ioService.properties();
+    }
+
+    @Override
+    public void handlerAdded() {
+        initSrcBuffer(PROTOCOL_LENGTH);
+    }
+
+    @Override
+    public HandlerStatus onRead() {
+        src.flip();
+
+        try {
+            if (src.remaining() < PROTOCOL_LENGTH) {
+                // The protocol has not yet been fully received.
+                return CLEAN;
+            }
+
+            String protocol = loadProtocol();
+
+            if (CLUSTER.equals(protocol)) {
+                initChannelForCluster();
+            } else if (CLIENT_BINARY_NEW.equals(protocol)) {
+                initChannelForClient();
+            } else {
+                // text doesn't have a protocol; anything that isn't cluster/client protocol will be interpreted as txt.
+                initChannelForText(protocol);
+            }
+
+            if (!channel.isClientMode()) {
+                protocolEncoder.signalProtocolEstablished(protocol);
+            }
+
+            return CLEAN;
+        } finally {
+            compactOrClear(src);
+        }
+    }
+
+    private String loadProtocol() {
+        byte[] protocolBytes = new byte[PROTOCOL_LENGTH];
+        src.get(protocolBytes);
+        return bytesToString(protocolBytes);
+    }
+
+    private void initChannelForCluster() {
+        channel.config()
+                .setOption(SO_SNDBUF, props.getInteger(SOCKET_RECEIVE_BUFFER_SIZE) * KILO_BYTE);
+
+        TcpIpConnection connection = (TcpIpConnection) channel.attributeMap().get(TcpIpConnection.class);
+        connection.setType(MEMBER);
+        channel.inboundPipeline().replace(this, ioService.createMemberInboundHandlers(connection));
+    }
+
+    private void initChannelForClient() {
+        channel.config()
+                .setOption(SO_RCVBUF, KILO_BYTE * clientRcvBuf())
+                // clients dont support direct buffers
+                .setOption(DIRECT_BUF, false);
+
+        TcpIpConnection connection = (TcpIpConnection) channel.attributeMap().get(TcpIpConnection.class);
+        channel.inboundPipeline().replace(this, new ClientMessageDecoder(connection, ioService.getClientEngine()));
+    }
+
+    private void initChannelForText(String protocol) {
+        channel.config()
+                .setOption(SO_RCVBUF, clientRcvBuf());
+
+        TcpIpConnection connection = (TcpIpConnection) channel.attributeMap().get(TcpIpConnection.class);
+        connection.getConnectionManager().incrementTextConnections();
+
+        TextEncoder encoder = new TextEncoder(connection);
+
+        channel.attributeMap().put(TextEncoder.TEXT_ENCODER, encoder);
+
+        TextDecoder decoder = new TextDecoder(connection, encoder);
+        decoder.src(newByteBuffer(channel.config().getOption(SO_RCVBUF), channel.config().getOption(DIRECT_BUF)));
```

Comment:
There can't be more bytes because the protocol buffer will only take 3 bytes.

This whole restoring is kinda shitty; but that is unfortunately how rest is implemented in our case.